src/device_selector.cc

```diff
@@ -0,0 +1,250 @@
+/*!
+ * Copyright 2014-2022 by Contributors
+ * \file device_selector.cc
+ */
+
+#include <xgboost/device_selector.h>
+#include <dmlc/registry.h>
+#include <xgboost/generic_parameters.h>
+#include <xgboost/json.h>
+
+namespace dmlc {
+DMLC_REGISTRY_ENABLE(::xgboost::KernelsReg);
+DMLC_REGISTRY_ENABLE(::xgboost::DeviceReg);
+}  // namespace dmlc
+
+namespace xgboost {
+
+DeviceSelector::Specification& DeviceSelector::Fit() {
+  return fit_;
+}
+
+DeviceSelector::Specification& DeviceSelector::Predict() {
+  return predict_;
+}
+
+DeviceSelector::Specification DeviceSelector::Fit() const {
+  return fit_;
+}
+
+DeviceSelector::Specification DeviceSelector::Predict() const {
+  return predict_;
+}
+
+DeviceSelector::DeviceSelector() {
+  std::stringstream ss;
+  ss << fit_.Prefix() << fit_ << "; " << predict_.Prefix() << predict_;
+  user_input_ = ss.str();
+}
+
+void DeviceSelector::Init(const std::string& user_input_device_selector) {
+  int fit_position = user_input_device_selector.find(fit_.Prefix());
```

Comment:
Do you think it's appropriate that we don't distinguish between predict and fit? Whatever device the user has specified, we will use it everywhere.

Reply:
We will remove `gpu_predictor` and `gpu_hist` (hopefully in this release) as documented in https://github.com/dmlc/xgboost/issues/7308 . The expected result is we will have only one (global) parameter `device` to control the dispatch:

``` python
with xgb.config_context(device="sycl:0"):
   booster.predict(X)
```

Reply:
Oh, I see. 
If you don't plan supporting different devices for fitting and prediction, this feature is inappropriate. Fortunately, it can easily be reduced to the uniform device descriptor for both stages.

Reply:
That avoids some internal conflicts, it's difficult to configure the states with the current design. We have been working on using https://github.com/dmlc/xgboost/blob/master/include/xgboost/generic_parameters.h as the context object for XGBoost. Maybe we can integrate the device selector in this PR with this context object?

Reply:
Sounds good

Reply:
What help do you need on this issue from my side?

Reply:
I will make some progress on setting up the interface and keep you posted. Thank you for working on it.

Reply:
Hi @trivialfis , 
are there any progress in this direction? May be some help from our side can be useful?

Reply:
@razdoburdin I have had some experiments on this recently, the problem is distributed environments and multi-threaded environments (like python async). We need to share the device index between all workers and all threads, which needs some synchronization strategies.

We don't need any synchronization if the device id is limited to booster as a local variable. But if we were to extend it to DMatrix as well (for constructing DMatrix from various sources of data), then the issue becomes a headache.

Reply:
Hi @trivialfis , have we a chance to implement such or similar concept in xgboost 2.0? Maybe you need some help in it?

Reply:
Yes, I'm still planning it as a major breaking change for 2.0. Got distracted away in the 1.7 by the new pyspark interface. Expect some progress next month. Sorry for the slow update.